src/Storages/StorageMerge.cpp

```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
     return builder;
 }
 
+ReadFromMerge::RowPolicyData::RowPolicyData(RowPolicyFilterPtr row_policy_filter_ptr_,
+    std::shared_ptr<DB::IStorage> storage,
+    ContextPtr local_context)
+    : row_policy_filter_ptr(row_policy_filter_ptr_)
+{
+    assert(row_policy_filter_ptr_);
+
+    ASTPtr expr = row_policy_filter_ptr->expression;
+
+    auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+    auto storage_columns = storage_metadata_snapshot->getColumns();
+    auto needed_columns = storage_columns.getAllPhysical();
+
+    auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
+    auto expression_analyzer = ExpressionAnalyzer{expr, syntax_result, local_context};
+
+    actions_dag = expression_analyzer.getActionsDAG(false /* add_aliases */, false /* project_result */);
+    filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+        ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+    const auto & required_columns = filter_actions->getRequiredColumnsWithTypes();
+    const auto & sample_block_columns = filter_actions->getSampleBlock().getNamesAndTypesList();
+
+    NamesAndTypesList added, deleted;
+    sample_block_columns.getDifference(required_columns, added, deleted);
+    if (!deleted.empty() || added.size() != 1)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Cannot determine row level filter");
+    }
+
+    filter_column_name = added.getNames().front();
+}
+
+void ReadFromMerge::RowPolicyData::extendNames(Names & names)
+{
+    std::sort(names.begin(), names.end());
+    NameSet added_names;
+
+    for (const auto & req_column : filter_actions->getRequiredColumns())
+    {
+        if (!std::binary_search(names.begin(), names.end(), req_column))
+        {
+            added_names.insert(req_column);
+        }
+    }
+    if (!added_names.empty())
+    {
+        std::copy(added_names.begin(), added_names.end(), std::back_inserter(names));
+    }
+}
+
+void ReadFromMerge::RowPolicyData::addStorageFilter(SourceStepWithFilter * step)
+{
+    assert(row_policy_filter_ptr);
```

Comment:
Not really necessary because you've already checked that in the constructor

Reply:
> Not really necessary because you've already checked that in the constructor

That's an idea of assert.
We state that we rely on something that is provided by another code.

Reply:
We usually don't assert pointers to be not null, because if they happen to be null the first access by that pointer will stop the program immediately with a fatal error - and it's almost the same as if the program fails on `assert()`. So it doesn't make much sense in checking pointers with `assert()`.